keymanager/src/runtime/methods.rs

```diff
@@ -54,35 +57,84 @@ const MAX_EPHEMERAL_KEY_AGE: EpochTime = 10;
 /// of blocks lower than the height of the latest trust root.
 const MAX_FRESH_HEIGHT_AGE: u64 = 50;
 
-/// The size of an encrypted ephemeral secret.
-const EPHEMERAL_SECRET_STORAGE_SIZE: usize = 32 + TAG_SIZE + NONCE_SIZE;
-
 /// Initialize the Kdf.
 pub fn init_kdf(ctx: &mut RpcContext, req: &InitRequest) -> Result<SignedInitResponse> {
+    let rctx = runtime_context!(ctx, KmContext);
+    let runtime_id = rctx.runtime_id;
+    let storage = ctx.untrusted_local_storage;
+
     let policy_checksum = Policy::global().init(ctx, &req.policy)?;
-    Kdf::global().init(req, ctx, policy_checksum)
+
+    let kdf = Kdf::global();
+    kdf.set_runtime_id(runtime_id)?;
+
+    // Load master secrets if not already loaded. It is intended that manual intervention
+    // by the operator is required to remove/alter persisted secrets.
+    kdf.load_master_secrets(storage, &runtime_id)?;
+
+    // Replicate the missing secrets.
+    let next_generation = kdf.next_generation();
+    if !req.checksum.is_empty() && next_generation <= req.generation {
+        let nodes = key_manager_nodes(ctx, runtime_id)?;
+        for generation in next_generation..=req.generation {
+            let secret = fetch_master_secret(ctx, generation, &nodes)?;
+            kdf.add_master_secret(storage, &runtime_id, secret, generation)?;
+        }
+    }
+
+    // TODO: Allow enclaves to generate master secrets for all generations.
```

Comment:
Why? To support starting with a non-zero generation?

Reply:
No, Maybe the comment is not clear. What I meant was that this part is legacy code which should be removed (in the last part) and replaced with the new process: 
- generate and propose the next master secret 
- decrypt and store proposed secret
- if accepted, transition to the next secret